proxy/src/proxy.rs

```diff
@@ -561,15 +567,13 @@ impl<S: AsyncRead + AsyncWrite + Unpin> Client<'_, S> {
             application_name: params.get("application_name"),
         };
 
-        let auth_result = async {
-            // `&mut stream` doesn't let us merge those 2 lines.
-            let res = creds
-                .authenticate(&extra, &mut stream, allow_cleartext)
-                .await;
-
-            async { res }.or_else(|e| stream.throw_error(e)).await
-        }
-        .await?;
+        let auth_result = match creds
+            .authenticate(&extra, &mut stream, allow_cleartext)
+            .await
+        {
+            Err(e) => return stream.throw_error(e).await,
+            Ok(auth_result) => auth_result,
```

Comment:
```suggestion
            Ok(auth_result) => auth_result,
            Err(e) => return stream.throw_error(e).await,
```

Maybe keep the consistent order...?